### PR TITLE
Remove from disk menu item

### PR DESCRIPTION
**Pull request by WaylonR**
Method to remove a track from disk, and optionally hides it from the library, moving it to 'hidden' tracks. Will go to 'missing' after a rescan.

### CODE REVIEW

**Comment by daschuer** [comment]
Thank you for the PR.
This is a modal message Box stopping the rest of Mixxx, right?
This should be replaced with one that is non-modal. Maybe you could reuse this.
https://github.com/mixxxdj/mixxx/blob/1c29f2a19d3f4c3864fed2e7c86eddfab8d6ebd3/src/recording/recordingmanager.cpp#L297

File: src/widget/wtracktableview.cpp
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: The Remove word is already used. The file does not necessary stored on a disk.
How about "Delete from file system"

**Comment by daschuer** [comment]
A "Also hide track in library" is hard to understand. 
...

This gives the chance to optimize the feature for the underlying use cases.
Mine are these: 
* I have a duplicated track an I like to keep the better one
* I like to delete the track because I hate it.
* I like to delete the track, because it is broken, but I like to replace it later.

What are you use cases? Can we reflect that in the dialog?

**Comment by daschuer** [comment]
The message Box should show also the path from the file to delete.

**Comment by WaylonR** [comment]
I am unsure how to make "Also hide track" checkbox, easier to understand. Please, feel free to write me a message.

**Comment by WaylonR** [comment]
The 'remove from disk' can remove multiple files at a time.. so, each file should be in a list that gets displayed in the message?

**Comment by daschuer** [comment]
Yes.

**Comment by daschuer** [comment]
First we have to clarify the use cases.
Is my list complete?

**Comment by WaylonR** [comment]
Yes, I think you've got a complete list.

**Comment by WaylonR** [comment]
As for model, it does not stop mixxx. It prevents access to the UI, yes, but mixxx itself does not stop. Nor the UI, or the music. The only thing that the model dialog does, is prevents a mouse/keyboard event inside mixxx. And to be fair, it is asking an important question.

**Comment by ronso0** [comment]
Thanks for picking this up!
@daschuer I think your list is complete.

> The message Box should show also the path from the file to delete.

Please show the complete file name & path so users can be sure to remove the track from the right subfolder (.../music/AlbumXY-copy/...) and catch the file extension as well as individual appendixes like 'supertrack_preview.mp3'

File: src/widget/wtracktableview.cpp
Comment (rryan) [inline src/widget/wtracktableview.cpp]: Let's handle failure to remove -- keep a list of filenames that we couldn't remove and let the user know somehow...

File: src/widget/wtracktableview.cpp
Comment (rryan) [inline src/widget/wtracktableview.cpp]: The list could contain duplicates paths -- let's de-dupe first.

File: src/widget/wtracktableview.cpp
Comment (rryan) [inline src/widget/wtracktableview.cpp]: shouldn't this be a purge? if the file is gone we should probably permanently remove the metadata too

**Comment by rryan** [comment]
It'd be nice to add some unit tests for this feature, since it's quite dangerous.

File: src/widget/wtracktableview.cpp
Comment (daschuer) [reply src/widget/wtracktableview.cpp]: That depends on the desired use case.
I will try to propose texts to distinguish it.

**Comment by daschuer** [comment]
Something like this:

[ ] Delete and replace crates and playlist entried with a successor  track.
[ ] Purge, emove Track from Playlist crates and History.
[ ] Delete track and mark as missing. 
[ ] Delete track and keep metadat as hidden track in history. 

Not perfect, but an idea.

File: src/widget/wtracktableview.cpp
Comment (WaylonR) [reply src/widget/wtracktableview.cpp]: So, two lists, a to do list with the dialog, and a dialog that shows any error, with a list of files which did not get deleted.

File: src/widget/wtracktableview.cpp
Comment (WaylonR) [reply src/widget/wtracktableview.cpp]: Tried to purge using trackModel->purgeTracks but.. it didn't work. not sure why. Even did hideTracks then purgeTracks. They stayed in hidden. So.. kinda stumped.

**Comment by daschuer** [comment]
We have "hide" this is original for hiding podcasts or other non music tracks. The track is moved to hidden tracks. This prevents the track from appearing again after library scan. 

Than we have Purge, which removes the track from library as if it was never there.

The problematic thing here is the history. 
The history should not be altered because of a changed library.

Looking at the wall of possible option, it can be anoying if a user just want to delete a bad track that was just added.

We can think about reverting the logic.
We may check if the track is member of any playlist and crate. And if not, we can delete the file and Purge the metadata, whithout asking anything else then a simple confirm.

If this is not the case we reject the delete:
"The track is used in .... "

**Comment by daschuer** [comment]
Alternative, for the case crate an playlist entried are found:

[ ] Just delete track file and keep references for a  successor track.
[ ] Purge, remove Track from Playlist crates and History.
[ ] Hide, remove Track from Playlist crates and keep it in History.

File: src/widget/wtracktableview.cpp
Comment (WaylonR) [reply src/widget/wtracktableview.cpp]: Good point.

why not just 'Delete File'?

File: src/widget/wtracktableview.cpp
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: This should be a simple and explicit question, no frills. The text is too complicated for confirming an irreversible action.

File: src/widget/wtracktableview.cpp
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: I suppose this action is not used very often and at least not while performing. It doesn't deserve enlarging and cluttering the top level layer of the context menu.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/widget/wtracktableview.cpp
Comment (Be-ing) [reply src/widget/wtracktableview.cpp]: I'd prefer "Delete file". I'm more concerned about users not understanding what a "file system" or "disk" is than a "file".

File: src/widget/wtracktableview.cpp
Comment (Be-ing) [reply src/widget/wtracktableview.cpp]: How about: "Are you sure you want to permanently delete this file?"/"Are you sure you want to permanently delete these files?" It would be nice if the code could distinguish between having a single file selected and multiple files selected to show the appropriate string. Also, I'd like to see the paths of files before deleting them.

File: src/widget/wtracktableview.cpp
Comment (Be-ing) [inline src/widget/wtracktableview.cpp]: Style nitpicks:
  * Delete extra blank lines.
  * Put brackets on same line as if/else statements:
```
if (condition == true) {
    foo();
} else {
   bar();
}
```

**Comment by Be-ing** [comment]
There is now a merge conflict with master.

**Comment by ronso0** [comment]
This works well, thank you!

Any special cases that deserve further testing?

**Comment by ronso0** [comment]
Is it really to late to merge this into 2.3?
It's improving the preparation workflow A LOT!

**Comment by uklotzde** [comment]
The requested changes have not been resolved, yet. This action performs a destructive task that must be save to use.

**Comment by uklotzde** [comment]
It is still unclear if the corresponding track of a deleted file should be purged from the library.

**Comment by ronso0** [comment]
> It is still unclear if the corresponding track of a deleted file should be purged from the library.

I go along with the choices @daschuer listed above:
[  ] Just delete track file and keep references for a successor track.
[  ] Hide, remove Track from Playlist crates and keep it in History.
[  ] Purge, remove Track from Playlist crates and History.

I imagine this to work like OS file copy/move dialogs: ask for a choice for every file that should be deleted, with a checkbox [  ] Apply choice to all files in queue

File: src/widget/wtracktableview.cpp
Comment (ronso0) [reply src/widget/wtracktableview.cpp]: What about a `Hide / Delete` submenu?
Could this hold items that distinguish between Hide, Delete but keep in history, Delete & Purge etc. or is it better to decide that per file in the dialog window?

**Comment by WaylonR** [comment]
So, a per file dialog, unless "apply to all" is checked at some stage.

> [ ] Just delete track file and keep references for a successor track.
> [ ] Hide, remove Track from Playlist crates and keep it in History.
> [ ] Purge, remove Track from Playlist crates and History.
What code does each of these use?
Just delete, and keep all references, would be:
            QString  filenameWithPath = trackModel->getTrackLocation(index);
            QFile file (filenameWithPath);
            file.remove();
This is fine, until the next refresh is done, which hides the file, as missing, anyway.

> [ ] Apply to all files in queue`

**Comment by ronso0** [comment]
For a while I run my personal branch with this PR and it works well.
What I noticed when deleting Mixxx recordings is that related cue files remain (same filename with .cue suffix). I'd love those to be removed as well.
What about letting Mixxx look for cue files in audio file's directory and ask if it should delete them?

**Comment by WaylonR** [comment]
Thank you for using this PR. What defines a recording? how easy is it to identify one?Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: ronso0 <notifications@github.com> Date: 5/05/19  11:53 AM  (GMT+12:00) To: mixxxdj/mixxx <mixxx@noreply.github.com> Cc: WaylonR <waylon@winged7.co.nz>, Author <author@noreply.github.com> Subject: Re: [mixxxdj/mixxx] Remove from disk menu item (#1737) For a while I run my personal branch with this PR and it works well.
What I noticed when deleting Mixxx recordings is that related cue files remain (same filename with .cue suffix). I'd love those to be removed as well.
What about letting Mixxx look for cue files in audio file's directory and ask if it should delete them?

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/mixxxdj/mixxx/pull/1737#issuecomment-489374325",
"url": "https://github.com/mixxxdj/mixxx/pull/1737#issuecomment-489374325",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by ronso0** [comment]
> What defines a recording? how easy is it to identify one?

I'm not necessarily referring to Mixxx recordings. There might also be external files from a mobile recorder or another app, also files might be quite small like aborted Mixxx recordings, or field recording snippets.
When removing **anyAudioFile**.mp3/wav/... just look if there's an associated **anyAudioFile**.cue in the same folder and propose to delete it as well.

File: src/widget/wtracktableview.cpp
Comment (ronso0) [reply src/widget/wtracktableview.cpp]: Could we find out the reason deletion failed?
File doesn't exist? no write access/rights?

The 'Success/Error' window could show either "OK" or a list:
- inexistant files
- failed deletions (try again?)

I realize it might be a not-so-common use case and hard to accomplish, but in case of deletion fails (partly) and the error window was closed, could the track view select the 'failed' tracks?

**Comment by ronso0** [comment]
merge conflicts are resolved here https://github.com/WaylonR/mixxx/pull/2

**Comment by ronso0** [comment]
@WaylonR would you mind merging https://github.com/WaylonR/mixxx/pull/2 ?
After that, do you think you can also take care of the merge conflicts with master?
Recently the track menu was separated in #2612 and is now also available as right-click menu on track property labels in decks.

**Comment by nwautier** [comment]
Sorry if this isn't the way you like to hear it, but BUMP.  This is a killer feature I'd love to have yesterday.  I'm not really comfortable diving into the code on a project of this scale, but if you want to bounce UI ideas or test cases off of me, feel free.

**Comment by Be-ing** [comment]
It's not so intimidating to get into such a big project once you have an IDE set up to navigate the code easily. I felt intimidated for a while until I got KDevelop setup. Now that we are using CMake on the master branch, most C++ IDEs are easy to set up.

File: src/widget/wtracktableview.cpp
Comment (nwautier) [reply src/widget/wtracktableview.cpp:675]: How about "This will move one or more files to the recycle bin, are you sure you want to continue?"

I feel that some of the less computer literate think of the Recycle Bin as a keyword almost.

File: src/widget/wtracktableview.cpp
Comment (nwautier) [reply src/widget/wtracktableview.cpp:1102]: I'm not too worried about cluttering context menu...  How often do you really right-click while performing?  I do most of my file selection with a hardware controller and never touch a mouse anyways.  I would not want another dialog box for each file...  If I selected a range, then hit purge, then purge all I selected...

File: src/widget/wtracktableview.cpp
Comment (nwautier) [reply src/widget/wtracktableview.cpp:451]: I like "Delete File" or "Move To Trash"

**Comment by WaylonR** [comment]
Looking to resolve this, I can't find the Hide action anymore? Only the Purge action.

File: src/widget/wtracktableview.cpp
Comment (WaylonR) [reply src/widget/wtracktableview.cpp:696]: It is a permanent deletion, from the disk. It doesn't go to the recycle bin, it's simply unlinked? @uklotzde ,  do you have a suggestion as to wording? This is a pretty scary action.

**Comment by ronso0** [comment]
Converted to Draft as there are a few open UI issues before we can implement it.

**Comment by ronso0** [comment]
I'll pick this up soonish because I miss it in my personal builds. I'll only try to restore the current state, not sure if I find the time to add the dialogs we talked about.

**Comment by ronso0** [comment]
I've ported this to 2.3 https://github.com/ronso0/mixxx/tree/remove-from-disk-2.3 to have it available in my 2.3 builds I use live. Shouldn't be too hard to apply it to master.
Respective PR to my 2.3 branch is https://github.com/ronso0/mixxx/pull/11
I added a simple file list to the message box:
image

**Comment by Holzhaus** [comment]
There are merge conflicts now.

**Comment by ronso0** [comment]
@WaylonR 
If you're still around, please spin back before merging b979bbe which is unrelated to this branch.

**Comment by ronso0** [comment]
> @WaylonR
> If you're still around, please spin back before merging [b979bbe](https://github.com/mixxxdj/mixxx/commit/b979bbedbbdfeb347c36290c55ff64720735c2d2) which is unrelated to this branch.

Sorry fo rthe noise, it was just the PR hover peview that links to https://github.com/mixxxdj/mixxx/pull/2 instead of my repo...

**Comment by ronso0** [comment]
I'll take care of the conflicts.

**Comment by ronso0** [comment]
Looking closer I think it's way easier to start over. I recommend to close this.
The trackmenu was moved out the library to wtrackmenu, so let's add the Remove action there.

**Comment by WaylonR** [comment]
please @notify me when this is picked up in another PR. might get around to doing it again, myself. I've been using a artifact from an old main branch compile, and been missing the feature.

**Comment by ronso0** [comment]
I just spent my time after dinner to rebuild this onto mas.. main.
wasn't too hard, works well in Tracks & Crates.
dirty branch is https://github.com/ronso0/mixxx/tree/remove-from-disk

I'll open a draft PR during the weekend and we can discuss where else it is needed, and how to tweak the user dialog.

**Comment by ronso0** [comment]
> I just spent my time after dinner to rebuild this onto mas.. main.

...because I missed it a lot since having rebuil my Live branch onto main.
Essential feature for my 'clean up the audio inbox'   workflow!

**Comment by ronso0** [comment]
@WaylonR  Here you go #3212

**Pull request closed without merging**
